SQS message sent to queue when book record changed

gaurijo opened the pull request:
This PR is for [issue 7](https://github.com/orgs/medusa-project/projects/6/views/1?pane=issue&itemId=35965754). This was tested in development mode similarly to [how the service check automation was tested in development](https://github.com/medusa-project/book-tracker/issues/6#issuecomment-1828668964).

Summary of Changes:

- `SQS `was created via `Terraform` in [aws demo repo](https://github.com/UIUCLibrary/aws-book-tracker-demo-service) - _will need to be done for setting up a prod sqs queue also_
- The logic for the service checks was expanded upon so that `after each service check a SQS message will be sent with information on records that were updated. `
- Eventually this can be consumed by the `Digital Library Search Gateway`. 

Screenshots below show how the sqs message is triggered after a service check runs:

<img width="1436" alt="Screenshot 2024-01-02 at 2 14 29 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/0dcce401-b12f-4a22-83f4-145dfcbf5515">

<img width="1223" alt="Screenshot 2024-01-02 at 2 57 44 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/3deb6a69-282c-4189-a200-5ea8a6f54ad5">

<img width="820" alt="Screenshot 2024-01-02 at 2 08 25 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/090b6f32-21ee-4003-a83b-b09acc2b3a29">

_Note_: I had written a test for the `sending of the sqs message`, which did pass locally, but threw an error in `GitHub Actions`. I'm working on figuring out how exactly to resolve it (at first I thought it was because the `edited credentials yml `hadn't been staged for commit, but that didn't change the error), so I commented the test out for now: 

```
book-tracker-test_1  | Error:
book-tracker-test_1  | RecordSourceTest#test_sends_a_message_sqs:
book-tracker-test_1  | Aws::Errors::MissingCredentialsError: unable to sign request without credentials set
book-tracker-test_1  |     test/models/record_source_test.rb:11:in `block in <class:RecordSourceTest>'
book-tracker-test_1  | 
book-tracker-test_1  | bin/rails test test/models/record_source_test.rb:6
```

gaurijo committed:
Add aws-sdk-sts and sqs gems

gaurijo committed:
call on SQS client with book tracker queue name and send message with updated records after service checks run

gaurijo committed:
Test: sqs queue is able to send a message as expected

gaurijo committed:
aws missing credentials error

gaurijo committed:
comment out sqs test message for now

adolski commented:
This is a good first stab. What we want to do, though, instead of sending a summary message at the end of the import, is send one message per imported book.

I can see how this might have been awkward to do based on how the code was designed, so I've pushed a commit to `develop` that should make it easier. It redesigns `Book.params_from_marcxml_record()` as `Book.from_marcxml_record()` to return a Book instance instead of a hash. That way, the argument that is passed to `RecordSource.upsert()` is an array of Book objects, rather than an array of hashes containing book properties.

So the changes needed could be something like this:

#### In Book:
```ruby
##
# @return [Hash]
#
def as_message()
  # I guess we don't know exactly what we want the contents
  # of the message to be yet, but for starters, we could just
  # have this wrap as_json().
end

##
# @param message [Hash] Hash returned from {as_message}.
#
def send_message(message)
  # your message-sending code goes here
end
```

#### In RecordSource:
```ruby
def upsert(batch)
  Rails.logger.debug("RecordSource.upsert(): upserting #{batch.length} records")
  Book.bulk_upsert(batch)
  # new code to send messages
  batch.each do |book|
    message = book.as_message
    book.send_message(message)
  end
ensure
  batch.clear
end
```

Does this make sense?

gaurijo commented:
Ohhh, I didn't realize we wanted the message sent after each record updated and not a summary of updated changes. What you laid out makes sense, thank you. I'll let you know if I run into any blockers.

gaurijo committed:
Merge branch 'develop' into SQS

gaurijo committed:
New instance methods for creating sqs message as json, and sending that message through the SQS

gaurijo committed:
Update pvt method .upsert(batch) to iterate through batch of new/updated book objects and call on the send_message method from the Book model to feed through SQS

gaurijo committed:
Test: verifies that message is sent as JSON through SQS upon new/updated book object

gaurijo committed:
Remove record source test file (may add back in later)

gaurijo committed:
Update: comment out sending sqs message due to aws credentials error in CI pipeline

gaurijo commented:
I just pushed up some more commits based on your suggested changes.  I ran an `import in development` (which ran super slow) to trigger the Book model `upsert`method. Here is how the SQS message looks like now with these changes: 

<img width="1289" alt="Screenshot 2024-01-03 at 2 01 50 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/2133cd82-a7da-4380-98f9-194937e2d87a">

<img width="902" alt="Screenshot 2024-01-03 at 2 06 43 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/e4d69354-f846-4c92-8c03-d4997d9be7f1">

I included a test inside `book_test.rb` for the message sending action, and while it does pass locally, `GitHub Actions` keeps throwing an error due to a `missing aws credential.` I commented the test out while I figure out why it's doing that.

adolski commented on app/models/book.rb:
```diff
@@ -195,6 +195,20 @@ def self.from_marcxml_record(record:, key:)
   end
       
+  def as_message
+    {"SQS message" => "New or Updated Record for Book: #{self.title}"}
```
I know you haven't learned much about the Search Gateway yet, but basically its function is to harvest records from several different repositories (including the Book Tracker) in order to present a unified search interface. The harvesting currently involves periodically crawling each service. There is a button on its backend that triggers a harvest.

This feature is the first step in a new way to do harvesting. The idea is that a service will push a message to a queue when a record changes, and the Gateway will see it immediately and update its corresponding record. The Gateway won't have to crawl anymore.

(Crawling is not ideal for a few reasons: 1) there are periods between the crawls during which content can be out of date; 2) it places a lot of load on the services during crawls; 3) a lot of records get harvested that haven't changed since the last harvest, which is inefficient.)

In order for this new system to work, the message (returned by `as_message()` here) has to contain the book's properties, so that the Gateway won't have to hit a URL to get them, it can just read them right out of the message.

There is already a Book method which returns most/all of a Book's properties: `as_json()`. Therefore it would be a good first step to simply have `as_message()` wrap `as_json()` (i.e. call it and return what it returns).

gaurijo replied:
I see, so essentially we want the Gateway to consume the message `(as_message`) with a `book's properties` in json format. Since we have a method that puts it in json format already (`as_json()`), I can call on that inside the `as_message` method like so? 

```ruby
def as_json(options = { })
    # existing code here
end 

def as_message 
   self.as_json 
end 
```

Or does it need to be called on like `self.as_json.to_json` which turns it from a hash to actual json data.

adolski replied:
I'd have it return a hash, like `as_json()` does, like you've written.

adolski commented on Gemfile:
```diff
@@ -26,6 +26,8 @@ gem 'scars-bootstrap-theme', github: 'medusa-project/scars-bootstrap-theme',
 gem 'uglifier', '>= 1.3.0'
 gem 'uiuc_lib_ad', git: 'https://github.com/UIUCLibrary/uiuc_lib_ad.git'
+gem 'aws-sdk-sqs'
+gem 'aws-sdk-sts'
```
Do you know why the aws-sdk-sts gem is included? I think aws-sdk-sqs should work on its own, but I could be wrong.

gaurijo replied:
I included it based on looking at how sqs was set up in the `fixity` repo, but I can remove it

adolski commented on app/models/record_source.rb:
```diff
@@ -1,3 +1,4 @@
+require 'aws-sdk-sqs'
```
This line shouldn't be needed

adolski commented on config/credentials.yml.enc:
```diff
@@ -0,0 +1 @@
+IdYEf2PHDPsCMELoDBAx5T6Iq1tnod6BK1FNv21OY8KwanTFhemayfKAcrHNSTrT5Rjyal6eqMgLojOP0bE43SYst691k5wBUev6+qzkhGx3MGAXbNpr3CW11uTkYXUL0gadK/V3Pgc0u1e/gdJwiKWnfDOx5J1LIvm2s5S5mAPw/PQ4POByK4Ws7MFbwl+HvlLkccSMtRVxoHAlsLQQ//RppZk9BUnc3EiN8XYRIeXe4toG7DA3yGeJq9h/ooj7n7QGhlhJviIPpYTQAWc2tFPfOby4yvhPkF/0HpnSoVioC7xbLI/YmvAdIHlllkKjbaIXdrZwN39qIH7z5Lftt/2y0PfQdEGNKCdFdD9vd+UHrz85Oq+8rxB/RTcd/P1irwD8qYIplykHJZP ...
```
This file was probably created using `rails credentials:edit` without an environment flag at the end, like `-e demo` or `-e production`. We always need to have the flag. This file should be deleted.

adolski commented on test/models/book_test.rb:
```diff
@@ -165,4 +165,12 @@ class BookTest < ActiveSupport::TestCase
+  #   book.send_message(message)
+
+  #   assert_equal message, {"SQS message" => "New or Updated Record for Book: #{book.title}"}
+  # end
```
`send_message()` is probably going to be hard to test, but for now, please just add a test of `as_message()`.

gaurijo replied:
Test for `as_message` has been added now

gaurijo committed:
Remove aws-sdk-sts gem

gaurijo committed:
call on .as_json inside as_message method

gaurijo committed:
remove require statement at top of file

gaurijo committed:
update tests

gaurijo committed:
Delete config/credentials.yml.enc

gaurijo commented:
I made the suggested changes. Here is how a SQS message should look now when a record is changed:
<img width="1031" alt="Screenshot 2024-01-04 at 2 39 39 PM" src="https://github.com/medusa-project/book-tracker/assets/103534307/4d49ac31-dbe7-43f1-8e07-3365665b5c8f">

adolski commented on app/models/book.rb:
```diff
@@ -220,7 +220,20 @@ def as_json(options = { })
+  end
+  
+  def send_message(message)
+    sqs = Aws::SQS::Client.new(region: 'us-east-2')
```
Instead of hard-coding the region here, let's get it from the app configuration:

`Configuration.instance.storage[:books][:region]`

gaurijo replied:
@adolski When I call on the region from our app configuration, I see that the region is set to`'us-east-1'`, however our AWS console points to the region being `'us-east-2'`. Is there a reason for the difference?

```ruby
[3] pry(#<Book>)> Configuration.instance.storage
=> {"books"=>
  {"region"=>"us-east-1",
   "endpoint"=>"http://localhost:9000",
   "bucket"=>"book-tracker-books-test",
   "key_prefix"=>nil,
   "access_key_id"=>"minioadmin",
   "secret_access_key"=>"minioadmin"},
 "temp"=>
  {"region"=>"us-east-1", "endpoint"=>"http://localhost:9000", "bucket"=>"book-tracker-temp-test", "access_key_id"=>"minioadmin", "secret_access_key"=>"minioadmin"}}
```

adolski replied:
`Configuration.instance` is drawing from the current configuration, which in the development environment will be `config/credentials/development.yml`. If you open that file, you should see `storage.books.region` set to `us-east-1`. (But in development, it doesn't matter.)

In test environment (where the region also doesn't matter) it will be drawing from `test.yml` and in demo & production, `demo.yml.enc` and `production.yml.enc`, respectively (these are the two that you have you use `rails credentials:edit -e <environment>` to edit).

adolski commented on app/models/book.rb:
```diff
@@ -220,7 +220,20 @@ def as_json(options = { })
+  
+  def send_message(message)
+    sqs = Aws::SQS::Client.new(region: 'us-east-2')
+    queue_name = "book-tracker-demo"
```
Let's also store the queue name in the app configuration.

gaurijo committed:
Calls on config file region and queue name inside send_message()

gaurijo committed:
Adds aws key with queue_name inside config

gaurijo committed:
Merge branch 'SQS' of https://github.com/medusa-project/book-tracker into SQS

gaurijo commented:
Ok, thanks. I updated the `send_message()` method to now call on the configuration for both `region` and `queue_name.` I added those to the `template.yml` and also `edited the demo credentials` with the same, assuming that was necessary. I left the prod credentials alone for now.

adolski commented on config/credentials/template.yml:
```diff
@@ -54,6 +54,8 @@ medusa_admins_group: Library Medusa Admins
 # Configures the S3 storage.
 storage:
+  aws:
+    queue_name:        book-tracker-demo
```
Sorry, but could you use a different config key for the queue name? Instead of `storage.aws.queue_name`, something like `sqs.queue_name`. Let's move it out of the `storage` section because that's for the S3 storage.

gaurijo replied:
No problem, It's now moved out of the storage section and lives inside its own sqs section. The demo credentials were also updated with this.

```ruby
sqs:
   queue_name:          book-tracker-demo 
```

gaurijo committed:
Update config credentials so queue_name is not within storage

gaurijo committed:
Update send_message() with new config key naming

gaurijo committed:
Refactor send_message() to retrieve temporary credentials

gaurijo committed:
Test: use minitest mock to test sqs send_message

gaurijo committed:
Test: remove region variable

gaurijo commented:
So I tried implementing some `mocking` to see if I could test the `send_message `method, but unfortunately the test keeps failing in GitHub Actions. I'm no longer getting a `missing aws credentials error`, but now the error is more vague `(a missing '[]' for nil class error)`. If you notice anything off the bat in my code that needs to be fixed I can quickly implement it, but if it seems like it's not worth the hassle of getting this particular test to work, I'll just remove the latest commits and we can merge without this test.

gaurijo committed:
Update ci config credentials with sqs queue_name

gaurijo commented:
Final update - I realized I never updated the `ci.yml config` with the credentials for the sqs queue. Now all the tests are passing in Github Actions, so this should be ready for merging. Let me know if there is anything else that should be looked at here.

adolski commented on app/models/book.rb:
```diff
@@ -220,7 +220,24 @@ def as_json(options = { })
+    self.as_json
+  end
+  
+  def send_message(message)
```
One thing I forgot--we don't want to send any messages from the development environment, so we should guard against that here, by doing something like `return if Rails.env.development?`

gaurijo replied:
I added a line so it returns an error if it's in development

```ruby
  def send_message(message)
    return 'This feature does not work in development' if Rails.env.development?
    region = Configuration.instance.storage[:books][:region]
    queue_name = Configuration.instance.sqs[:queue_name]

     # rest of code 
  end
```

gaurijo committed:
Guard against sending sqs message when in development env

adolski commented:
@gaurijo Have you edited the demo Book Tracker's terraform script to create the demo SQS queue, and does the queue name have the word "demo" in it somewhere? We will need to have separate queues for demo and production.

gaurijo commented:
> @gaurijo Have you edited the demo Book Tracker's terraform script to create the demo SQS queue, and does the queue name have the word "demo" in it somewhere? We will need to have separate queues for demo and production.

Yes, that's been created [here](https://github.com/UIUCLibrary/aws-book-tracker-demo-service/blob/master/book-tracker-sqs.tf). Once this is merged I was going to work on setting up the production queue.
